src/include/miopen/env.hpp

```diff
@@ -31,127 +31,181 @@
+};
+
+template <>
+struct ParseEnvVal<std::string>
+{
+    static std::string go(const char* vp) { return std::string{vp}; }
+};
 
-// Return 0 if env is enabled else convert environment var to an int.
-// Supports hexadecimal with leading 0x or decimal
-inline uint64_t EnvvarValue(const char* name, uint64_t fallback = 0)
+template <typename T>
+struct EnvVar
 {
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
+private:
+    T value{};
+    bool is_default = true;
+
+public:
+    const T& GetValue() const { return value; }
+
+    bool IsDefault() const { return is_default; }
+
+    void UpdateValue(const T& val)
     {
-        return fallback;
+        is_default = false;
+        value      = val;
     }
-    else
+
+    explicit EnvVar(const char* const name, const T& def_val)
     {
-        return strtoull(value_env_p, nullptr, 0);
+        // NOLINTNEXTLINE (concurrency-mt-unsafe)
+        const char* vp = std::getenv(name);
+        if(vp != nullptr) // a value was provided
+        {
+            is_default = false;
+            if constexpr(std::is_same_v<T, bool>)
+            {
+                value = ParseEnvVal<bool>::go(vp);
+            }
+            else if constexpr(std::is_same_v<T, uint64_t>)
+            {
+                value = ParseEnvVal<uint64_t>::go(vp);
+            }
+            else if constexpr(std::is_same_v<T, std::string>)
+            {
+                value = ParseEnvVal<std::string>::go(vp);
+            }
+            else
+            {
+                value = ParseEnvVal<T>::go(vp); // should cause compile error
+            }
+        }
+        else // no value provided, use default value
+        {
+            value = def_val;
+        }
     }
+};
+
+} // end namespace internal
+
+// static inside function hides the variable and provides
+// thread-safety/locking
+#define MIOPEN_DECLARE_ENV_VAR(name, type, default_val)                    \
+    struct name                                                            \
+    {                                                                      \
+        using value_type = type;                                           \
+        static miopen::internal::EnvVar<type>& Ref()                       \
+        {                                                                  \
+            static miopen::internal::EnvVar<type> var{#name, default_val}; \
+            return var;                                                    \
+        }                                                                  \
+    };
+
+/// \todo the following functions should be renamed to either include the word Env
+/// or put inside a namespace 'env'. Right now we have a function named Value()
+/// that returns env var value as only 64-bit ints
+
+template <class EnvVar>
+inline std::string GetStringEnv(EnvVar)
```

Comment:
I missed this but we can promote the return type to a `const std::string&` reference because it's to a static variable, and thus avoid a copy.